### DISCUSSION TITLE
Could ESPNow.irecv() lose a transmission?

### DISCUSSION THREAD

**Discussion by mmliam** [micropython/micropython#11646] [General]
If code is currently executing a process external to the ESPNow.irecv() and an ESPnow transmission occurs could the transmission be lost, or does ESPNow.irecv() include interrupt processing to ensure receipt of all transmissions?

Thanks, MM

**Comment by davefes** [comment]
I have been working with the understanding that messages are queued in the rx_buff waiting for you to process them.  I generally only have one message in the queue.  Also, I think it only holds 250 bytes.

**Reply by mmliam** [reply]
Dave, The question is how and when are messages queued when the processor is executing code outside the ESPNow.irecv() process.

**Reply by davefes** [reply]
I hope, and here is where the designer has to chip-in ... that as long as the Wifi and e0 interfaces are active any incoming messages are saved in rx_buff ... until you process them.  I don't know what happens if the WiFi interface is down.

**Reply by mmliam** [reply]
The processor can only be executing code in one place at one time; I suspect the queuing of ESPNow messages is interrupt driven. Where ever the processor is executing it will be interrupted, executes code to queue a message, then the processor returns to continue where ever it was executing code. This is the only way I know of to prevent loss of ESPnow messaging. But, rather than speculating, I would like to hear from someone who really knows.

**Comment by glenn20** [comment]
tldr: The processing of incoming espnow packets is handled in an "interrupt-like" manner and events are saved in a buffer to be read out later with recv()/irecv(). So packets are not lost between calls to `recv()`.

The espressif wireless stack processes incoming wireless packets in high priority wifi tasks. When an espnow packet is received, a callback function in the espnow module is called from one of those high priority tasks to handle the packet. That function saves the packet in an internal buffer (if there is room in the buffer - if not, the packet is dropped).

You can change the buffer size with `e.config(rxbuf=xxxx)` (before `e.active(True)`) where xxxx is in bytes (see https://docs.micropython.org/en/latest/library/espnow.html#espnow.ESPNow.config). The default buffer size is enough for 2 full size packets (250bytes per packets plus overhead).

You can see if packets are being dropped with `e.stats()` (https://docs.micropython.org/en/latest/library/espnow.html#espnow.ESPNow.stats). If the packets are being dropped, you should try to clear the buffer more often (by calling `e.recv()/e.irecv()` more often) or make the buffer larger.

**Reply by mmliam** [reply]
Hi glenn,

Sorry I didn't see your response. I had since found this in the Espressif WiFi driver manual 

> The ‘Static Rx Buffer’ forms the hardware receiving list. Upon receiving a frame over the air, hardware writes the frame into the buffer and raises an interrupt to the CPU. Then, the Wi-Fi driver reads the content from the buffer and returns the buffer back to the list.


So indeed the wifi receive buffer is interrupt driven.
